src/Platform/Microsoft.Testing.Platform/OutputDevice/Console/AnsiTerminal.cs

```diff
@@ -0,0 +1,321 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System.Runtime.InteropServices;
+using System.Text;
+
+using Microsoft.Testing.Platform.Helpers;
+using Microsoft.Testing.Platform.Resources;
+
+namespace Microsoft.Testing.Platform.OutputDevice.Console;
+
+/// <summary>
+/// Console writer that is used when writing ANSI is allowed. It is capable of batching as many updates as possible and writing them at the end,
+/// because the terminal is responsible for rendering the colors and control codes.
+/// </summary>
+internal class AnsiTerminal : ITerminal
+{
+    private static readonly string[] KnownFileExtensions = new string[]
+    {
+        ".cs",
+        ".vb",
+        ".fs",
+        ".trx",
+        ".coverage",
+        ".log",
+        ".txt",
+    };
+
+    private readonly IConsole _console;
+    private readonly string? _baseDirectory;
+    private readonly bool _useBusyIndicator;
+    private readonly StringBuilder _stringBuilder = new();
+    private bool _isBatching;
+    private TestWorkerFrame _currentFrame = new(Array.Empty<TestWorker>(), 0, 0);
+
+    public AnsiTerminal(IConsole console, string? baseDirectory)
+    {
+        _console = console;
+        _baseDirectory = baseDirectory ?? Directory.GetCurrentDirectory();
+
+        // Output ansi code to get spinner on top of a terminal, to indicate in-progress task.
+        // https://github.com/dotnet/msbuild/issues/8958: iTerm2 treats ;9 code to post a notification instead, so disable progress reporting on Mac.
+        _useBusyIndicator = !RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+    }
+
+    public int Width
+        => _console.IsOutputRedirected ? int.MaxValue : _console.BufferWidth;
+
+    public int Height
+        => _console.IsOutputRedirected ? int.MaxValue : _console.BufferHeight;
+
+    public void Append(char value)
+    {
+        if (_isBatching)
+        {
+            _stringBuilder.Append(value);
+        }
+        else
+        {
+            _console.Write(value.ToString());
```

Comment:
The system console has an overload with `char`, so it's better to add it to `IConsole` and to use it here instead of forcing call to `ToString()`.